README
# stig-fix-el6

## About

DISA STIG Scripts to harden a system to the RHEL 6 STIG.

This script is currently being used for bare metal systems
see https://people.redhat.com/fcaviggi/stig-fix for the 
associated kickstart. Work is being done to merge in AWS,
Puppet, and Chef content.

These scripts will harden a system to specifications that
are based upon the the following hardening and specifications
provided by the following projects:

* [DISA RHEL 6 STIG V1 R2](http://iase.disa.mil/stigs/os/unix/red_hat.html)
* [NIST 800-53 (USGCB) Content for RHEL 5](http://usgcb.nist.gov/usgcb/rhel_content.html)
* [NSA SNAC Guide for Red Hat Enterprise Linux 5](http://www.nsa.gov/ia/_files/os/redhat/NSA_RHEL_5_GUIDE_v4.2.pdf)
* [Aqueduct Project](https://fedorahosted.org/aqueduct)
* [Tresys Certifiable Linux Integration Platform (CLIP)](http://oss.tresys.com/projects/clip]
     
## Content

Scripts are designed to run out of ```/opt/stig-fix/``` on a preferably fresh
installation of RHEL 6.4+ (RHEL 6.4 updated ```pam_lastlog.so``` to disable 
inactive users.)

* ```apply.sh``` - master script that runs scripts in cat1-cat4 and misc
* ```checkpoint.sh``` - checkpoint the current configuration so re-running apply.sh will not squash changes
* ```toggle_ipv6.sh``` - toggles IPv6 support, requires reboot (default is off)
* ```toggle_nousb.sh``` - toggles the 'nousb' kernel flag only
* ```toggle_udf.sh``` - toggles 'udf' mounting of DVDs (USGCB Blacklists udf) 
* ```toggle_usb.sh``` - toggles 'nousb' kernel flag and the mass storage kernel module
* ```config``` - Directory with some pre-STIGed configurations (auditd,iptables,system-auth-local,etc.) 
* ```scripts``` - Hardening Scripts
* ```misc``` - NSA SNAC, GNOME, and Other miscellenous lockdown scripts
* ```manual``` - Manaully run (There be dragons here)
* ```backups``` - Backup copy of modified files to compare and restore configurations

Run is logged in ```/var/log/stig-fix-YYYY-MM-DD.log```

## How-to
1. Hardening the BIOS configuration (Disable USB Booting, set administrative passwords, etc.)
2. Enable Drive Encryption (LUKS on your LVM VG or Hardware-based FIPS 140-2 compliant)
3. Partitioning the system to minimally include the following:
  * /boot
  * swap
  * /
  * /home
  * /tmp
  * /var
  * /var/log
  * /var/log/audit
  * /var/www (Optional, Web Servers)
  * /rhnsat (Optional, RHN Satellite 5.5 and earlier, 5.6 uses /var)
  * /opt (Optional, Commercial Applications)
4. Select a GRUB password (SHA512, using grub-crypt or during Install)
5. Install a minimal installation (include rsyslog, ntp, aide, scrub, vlock, screen, logwatch, openswan, openscap, openscap-utils, dracut-fips) [RPM pulls the base requirements]
6. Register a system with a patch server - either a disconnected RHN Satellite or local/web-based (preferably SSL) repository:
  * Kickstart can register a system with rhnreg_ks --activationkey="<KEY>"
  * With a disconnected RHN Satellite, be sure to include the proper /etc/sysconfig/rhn/up2date configuration and SSL Certificate.
  * For a local or web-based repository, create a .repo file in /etc/yum.repos.d/

    ```
    # vi /etc/yum.repos.d/rhel-dvd.repo
                                  
    [rhel-dvd]
    name=Red Hat Enterprise Linux $releasever - $basearch - DVD
    baseurl=file:///media/
    enabled=1
    gpgcheck=1
    gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release
    
    # mount /dev/dvd /media
    # yum clean all
    # yum update
    ```

  * For a repository, synchronize patches via a RHEL system connected to the Public RHN:
    * To synchronize a channel with only the newest packages, use the following command:

    ``` 
    # reposync -n -l -p <directory> -repoid=’<channel>’
    ```

    * To synchronize an entire channel, use the following command:

    ```
    # reposync -l -p <directory> -repoid=’<channel>’
    ```

    * After the packages have been downloaded into the directory in the previous step, create a repository with the RPMs.

    ```
    # createrepo <directory>
    ```

    * To export the directory, create an ISO image from the directory to burn to DVD:

    ```
    # mkisofs -RJ -o repo-export.iso <directory>
    ```

7. Include the stig-fix scirpts in the Kickstart and run the configuration '/sbin/stig-fix -q &> /dev/null' in the %post section  -OR- install the RPM afterwards and run manually (/sbin/stig-fix)
8. Configure NTP (/etc/ntp.conf) for the local time sources (or military ones) that are approved
9. Configure Syslog forwarding (/etc/rsyslog.conf) to a local rsyslog instance (or Splunk as the case may be)
10. Add users with root access to the 'wheel' group, for users with ssh access add them to the 'sshusers' group.
11. Review the manual scripts in (/opt/stig-fix/manual) - putting the kernel in FIPS 140-2 mode can cause various weaker authentication not to work (e.g. RHN Satellite is still using md5)
12. The following scripts have been included for convenience (RPM Symbolic Links to scripts):
  * /sbin/toggle_ipv6 (for IPv6 support - defualt is off)
  * /sbin/toggle_udf (for mounting DVDs)
  * /sbin/toggle_usb (for enabling and disabling USB storage)
  * /sbin/toggle_nousb (for enabling USB Keyboards that don't work with the 'nousb' kernel option)
13. Configure Firewalls/TCP_WRAPPERS:
  * Edit the iptables, ip6tables, hosts.allow, and hosts.deny as requried, copy any changes back to the ./config/ directory for iptables/ip6tables and modify cat2/gen006620.sh (TCP_WRAPPERS) to ensure changes are applied if the stig-fix command is run again.
14. Install site approved monitoring tools and virus scan. Examples include, but are not limited to, the following:
  * McAfee VSE for Linux
  * McAfee HBSS
  * ClamAV
  * HP OpenView
15. Create users:
  * Remote Access (no ssh without this group)

    ```
    # useradd -m -c "Remote User" -G sshusers remoteuser
    ```

  * System Administrator (SA)

    ```
    # useradd -m -c "System Administrator" -G sshusers,wheel admin
    ```

  * Audit Administrator (AA)

    ```
    # useradd -m -c "Audit Administrator" -G sshusers,isso auditor
    ```

  * (Optional) After adding SAs to the system, lock the root account:

    ```
    # passwd -l root
    ```

16. System Checkpoint - to prevent apply.sh from overwriting your configuration you should checkpoint your system. A checkpoint will backup the current configuration into the 'backup' directory and overwrite the configurations contained in the 'config' directory.

    ```
    # ./checkpoint.sh
    ```

  * Created the ability to make an RPM from the configuration using the following commands as root:
    * Install Requirements:

    ```
    # yum install rpmbuild make autoconf
    ```

    * Create RPM

    ```
    # make rpm
    ```

    * Clean up

    ```
    # make clean
    ```

## Project Work

* Map DISA STIG RHEL 5 GEN controls to DISA STIG RHEL 6 SRG and NIST 800-53 controls (each sub script has an echo block stating what GEN it applies to - adding the SRG and NIST controls will help security people to understand what was intended during the C&A process.
* Help verify the configurations against SSG OpenSCAP Content
* Add updates to with new security features in RHEL 6.5 and newer
* Merge in work for AWS, Puppet, Chef


CHANGELOG
* Mon Sep 16 2013 - fcaviggi (at) redhat.com
- Original Build
* Thu Sep 26 2013 - fcaviggi (at) redhat.com
- Fixed PAM Configuration
- Added 'sshusers' group for SSH limiting
- Added RPM dependancy on ntp
* Fri Sep 27 2013 - fcaviggi (at) redhat.com
- Added dependancy on openswan
* Mon Sep 30 2013 - fcaviggi (at) redhat.com
- Modified system-auth configuration for session pam_lastlog.so
- Standardized /etc/issue
- Modified system-auth configuration for password pam_cracklib.so
- Removed dependancy for dracut-fips, enabled downloading it in /opt/stig-fix/manual/gen000588-fips140-2.sh
* Tue Oct 01 2013 - fcaviggi (at) redhat.com
- Modified audit.rules to match SSG rules better
* Tue Oct 07 2013 - fcaviggi (at) redhat.com
- Fixed audit.rules to set system as immutable (-e 2) until reboot
* Tue Dec 13 2013 - fcaviggi (at) redhat.com
- Removed messagebus from disabled service due to X requirement
- Disabled ip6tables due to disablement of IPv6 in kernel configurations
* Tue Dec 17 2013 - fcaviggi (at) redhat.com
- Fixed pam_faillock for GDM use
* Tue Dec 18 2013 - fcaviggi (at) redhat.com
- Added netfs to disabled services list
- Updated sshd_configuration (Uncommented some defaults) and X11Forwarding to no by defualt
* Mon Feb 03 2014 - fcaviggi (at) redhat.com
- Fixed Audit rules
* Tue Apr 02 2014 - fcaviggi (at) redhat.com
- Updated documentation, added isso group for Auditors
* Wed May 14 2014 - fcaviggi (at) redhat.com
- Updated scripts to better comlpy with DISA STIG
* Wed May 15 2014 - fcaviggi (at) redhat.com
- Set BASE_DIR variable to '/opt/stig-fix' in apply.sh
* Wed May 26 2014 - fcaviggi (at) redhat.com
- Added toggle_ipv6.sh script and micellenous updates
* Sat Jul 05 2014 - fcaviggi (at) redhat.com
- Added setting in sshd_config to allow Kerberos passwords
- Added configuration file for logrotate
* Wed Jul 16 2014 - fcaviggi (at) redhat.com
- Fixed typo in audit.rules - privelged changed to privileged
* Wed Jul 23 2014 - fcaviggi (at) redhat.com
- Removed try_first_pass from pam_cracklib.so in system-auth-local and password-auth-local
- Fixed variable in cat2/gen0021000.sh
* Thu Jul 24 2014 - fcaviggi (at) redhat.com
- Fixed quotation issue around shells that was breaking sftp in cat2/gen002140.sh
* Fri Jul 25 2014 - fcaviggi (at) redhat.com
- Moved TCP_WRAPPERS configuration into 'config' directory, deleted cat2/gen006620.sh
- Added the ability to checkpoint configuration with checkpoint.sh script, aliased to /sbin/stig-fix-checkpoint
- Checkpoint prevents apply.sh from placing default configurations over a configured system
* Wed Aug 20 2014 - fcaviggi (at) redhat.com
- Added the ability to create RPM
* Sun Oct 12 2014 - fcaviggi (at) redhat.com
- Various fixes to scripts (password length cat2/GEN000580.sh)
- Fixed typo in etc/krb5.conf file
* Fri Dec 19 2014 - fcaviggi (at) redhat.com
- Merged SCAP Security Guide (SSG) Scripts - Thanks to Shawn Wells
- Updated directory structure for hardening scripts
* Sat Dec 20 2014 - fcaviggi (at) redhat.com
- Added misc/ctrlaltdel.sh


apply.sh
#!/bin/sh
# This script was written by Frank Caviggia, Red Hat Consulting
# Last update was 25 July 2014
# This script is NOT SUPPORTED by Red Hat Global Support Services.
# Please contact Josh Waldman for more information.
#
# Script: apply.sh (stig-fix)
# Description: RHEL 6 Hardening Script (Master Script)
# License: GPL (see COPYING)
# Copyright: Red Hat Consulting, Sep 2013
# Author: Frank Caviggia <fcaviggi (at) redhat.com>

# Determine the Path
function realpath() {
    local r=$1; local t=$(readlink $r)
    while [ $t ]; do
        r=$(cd $(dirname $r) && cd $(dirname $t) && pwd -P)/$(basename $t)
        t=$(readlink $r)
    done
    echo $r
}

# ENVIRONMENT VARIABLES
DATE=`date +%F`
ARCH=`uname -i`
VERSION='1.1'
BASE_DIR=`dirname $(realpath $0)`
BACKUP=$BASE_DIR/backups
CONFIG=$BASE_DIR/config
LOG=/var/log/stig-fix-$DATE.log

# Script Version
function version() {
	echo "Hardening Scripts for RHEL 6 (v.$VERSION)"
}

# USAGE STATEMENT
function usage() {
cat << EOF
usage: $0 [options]

  -v    Show version
  -h 	Show this message
  -q	Quiet output for scripting use

Hardening Scripts for RHEL 6 (v.$VERSION)

Applies Hardening Configurations to a system.

The scripts apply best-practice configurations based upon 
the following standards:

     DISA RHEL 6 STIG

     NIST 800-53 SCAP (USGCB)

     NSA SNAC Guide for Red Hat Enterprise Linux

     Aqueduct Project
     https://fedorahosted.org/aqueduct

     Tresys Certifiable Linux Integration Platform (CLIP)
     http://oss.tresys.com/projects/clip
EOF
}

while getopts ":vhq" OPTION; do
	case $OPTION in
		v)
			version
			exit 0
			;;
		h)
			usage
			exit 0
			;;
		q)
			QUIET=1
			;;
		?)
			echo "ERROR: Invalid Option Provided!"
			echo
			usage
			exit 1
			;;
	esac
done

if [ -z "$QUIET" ]; then
	echo -e "\033[3m\033[1mRed Hat Enterprise 6 Linux Hardening Scripts\033[0m\033[0m"
	echo
	echo "These scripts will harden a system to specifications that" 
	echo "are based upon the the following standards:"
	echo
	cat << EOF
     DISA RHEL 6 STIG

     NIST 800-53 SCAP (USGCB)

     NSA SNAC Guide for Red Hat Enterprise Linux

     Aqueduct Project
     https://fedorahosted.org/aqueduct

     Tresys Certifiable Linux Integration Platform (CLIP)
     http://oss.tresys.com/projects/clip
EOF
	echo
fi

# Check for root user
if [[ $EUID -ne 0 ]]; then
	if [ -z "$QUIET" ]; then
		echo
		tput setaf 1;echo -e "\033[1mPlease re-run this script as root!\033[0m";tput sgr0
	fi
	exit 1
fi

if [ -z "$QUIET" ]; then
	echo
	echo -e "\033[1mPlease snapshot or backup your system before running these scripts.\033[0m"
	echo
	echo -ne "\033[1mDo you want to continue?\033[0m [y/n]: "
	while read a; do
		case "$a" in
		y|Y)	break;;
		n|N)	exit 1;;
		*)	echo -n "[y/n]: ";;
		esac
	done
	echo
	echo -e "\033[1mStarting Configuration\033[0m"
fi

# CREATE LOG IF IT DOESN'T EXISIT
if [ ! -e $LOG ]; then
	touch $LOG
fi

echo "SCRIPT RUN: $(date)" >> $LOG
echo "Starting Configuration" >> $LOG

`rhn_check`
if [ $? -eq 0 ]; then
     echo '==================================================='
     echo ' Verifying RPM Requirements'
     echo '==================================================='
     yum -y install aide screen logwatch vlock openswan scrub
fi

# BACKUP ORIGINAL SYSTEM CONFIGURATIONS
if [ -z "$QUIET" ]; then
	echo -n "Back up current configuration... " | tee -a $LOG
else
	echo -n "Back up current configuration... " >> $LOG
fi

if [ ! -d $BACKUP ]; then
	mkdir -p $BACKUP
fi

if [ ! -f "$BACKUP/issue.orig" ]; then
	cp /etc/issue $BACKUP/issue.orig
fi

if [ ! -f "$BACKUP/issue.net.orig" ]; then
	cp /etc/issue.net $BACKUP/issue.net.orig
fi

if [ ! -f "$BACKUP/sysctl.conf.orig" ]; then
	cp /etc/sysctl.conf $BACKUP/sysctl.conf.orig
fi

if [ ! -f "$BACKUP/login.defs.orig" ]; then
	cp /etc/login.defs $BACKUP/login.defs.orig
fi

if [ ! -f "$BACKUP/audit.rules.orig" ]; then
	cp /etc/audit/audit.rules $BACKUP/audit.rules.orig
fi

if [ ! -f "$BACKUP/auditd.conf.orig" ]; then
	cp /etc/audit/auditd.conf $BACKUP/auditd.conf.orig
fi

if [ ! -f "$BACKUP/limits.conf.orig" ]; then
	cp /etc/security/limits.conf $BACKUP/limits.conf.orig
fi

if [ ! -f "$BACKUP/sshd_config.orig" ]; then
	cp /etc/ssh/sshd_config $BACKUP/sshd_config.orig
fi

if [ ! -f "$BACKUP/ssh_config.orig" ]; then
	cp /etc/ssh/ssh_config $BACKUP/ssh_config.orig
fi

if [ ! -f "$BACKUP/system-auth.pam.orig" ]; then
	cp /etc/pam.d/system-auth $BACKUP/system-auth.pam.orig
fi

if [ ! -f "$BACKUP/gnome-screensaver.orig" ]; then
	cp /etc/pam.d/gnome-screensaver $BACKUP/gnome-screensaver.orig
fi

if [ ! -f "$BACKUP/password-auth.pam.orig" ]; then
	cp /etc/pam.d/password-auth $BACKUP/pasword-auth.pam.orig
fi

if [ ! -f "$BACKUP/sudoers.orig" ]; then
	cp /etc/sudoers $BACKUP/sudoers.orig
fi

if [ ! -f "$BACKUP/smb.conf.orig" ]; then
	cp /etc/samba/smb.conf $BACKUP/smb.conf.orig
fi

if [ ! -f "$BACKUP/ntp.conf.orig" ]; then
	cp /etc/ntp.conf $BACKUP/ntp.conf.orig
fi

if [ ! -f "$BACKUP/krb5.conf.orig" ]; then
	cp /etc/krb5.conf $BACKUP/krb5.conf.orig
fi

if [ ! -f "$BACKUP/hosts.allow.orig" ]; then
	cp /etc/hosts.allow $BACKUP/hosts.allow.orig
fi

if [ ! -f "$BACKUP/hosts.deny.orig" ]; then
	cp /etc/hosts.deny $BACKUP/hosts.deny.orig
fi

if [ ! -f "$BACKUP/profile.orig" ]; then
	cp /etc/profile $BACKUP/profile.orig
fi

if [ ! -f "$BACKUP/logrotate.orig" ]; then
	cp /etc/logrotate.conf $BACKUP/logrotate.orig
fi

if [ ! -f "$BACKUP/iptables.orig" ]; then
	cp /etc/sysconfig/iptables /etc/sysconfig/iptables.orig
	cp /etc/sysconfig/iptables $BACKUP/iptables.orig
fi

if [ ! -f "$BACKUP/ip6tables.orig" ]; then
	cp /etc/sysconfig/ip6tables /etc/sysconfig/ip6tables.orig
	cp /etc/sysconfig/ip6tables $BACKUP/ip6tables.orig
fi

if [ -z "$QUIET" ]; then
	echo "Done." | tee -a $LOG
else
	echo "Done." >> $LOG
fi

# APPLYING DEFAULT SYSTEM CONFIGURATIONS
if [ -z "$QUIET" ]; then
	echo -n "Applying base configuration files... " | tee -a $LOG
else
	echo -n "Applying base configuration files... " >> $LOG
fi

# CHANGE DIRECTORY TO BASE DIR
cd $BASE_DIR

#### SSH CONIFIGURATION
cp -f ./config/sshd_config /etc/ssh/sshd_config
`ls /etc/ssh/ssh_host_* | grep -q key`
if [ $? -ne 0 ]; then
	/etc/init.d/sshd restart &> /dev/null
fi

#### KERNEL PARAMETERS
cp ./config/sysctl.conf /etc/sysctl.conf
/sbin/sysctl -p /etc/sysctl.conf

#### USER AND PASSWORD CONFIGURATIONS
cp -f ./config/limits.conf /etc/security/limits.conf
cp -f ./config/login.defs /etc/login.defs

#### EXAMPLE SAMBA CONFIGURATION
cp -f ./config/smb.conf /etc/samba/smb.conf

#### AUDITING RULES
cp -f ./config/auditd.conf /etc/audit/auditd.conf
if [ "$ARCH" == "x86_64" ]; then
	cp -f ./config/audit.rules /etc/audit/audit.rules
else
	grep -v 'b64' ./config/audit.rules > /etc/audit/rules.d/audit.rules
fi

# Remove RHEL 6.6 /etc/audit/rules.d directory
if [ -d /etc/audit/rules.d ]; then
	rm -rf /etc/audit/rules.d
fi

#### FIREWALL CONFIGURATIONS (IPV4/IPV6)
cp -f ./config/iptables /etc/sysconfig/iptables
cp -f ./config/ip6tables /etc/sysconfig/ip6tables

#### PAM CONFIGURATIONS
cp -f ./config/system-auth-local /etc/pam.d/system-auth-local
cp -f ./config/system-auth-local /etc/pam.d/system-auth-ac
ln -sf /etc/pam.d/system-auth-local /etc/pam.d/system-auth 
cp -f ./config/password-auth-local /etc/pam.d/password-auth-local
cp -f ./config/password-auth-local /etc/pam.d/password-auth-ac
ln -sf /etc/pam.d/password-auth-local /etc/pam.d/password-auth
cp -f ./config/gnome-screensaver /etc/pam.d/gnome-screensaver

##### SUDO CONFIGURATION (isso role, sudo access for wheel)
cp -f ./config/sudoers /etc/sudoers

##### LOGROTATE (DAILY)
cp -f ./config/logrotate.conf /etc/logrotate.conf

#### NTP CONFIGURATIONS
cp -f ./config/ntp.conf /etc/ntp.conf

#### KERBEROS CONFIGURATIONS
cp -f ./config/krb5.conf /etc/krb5.conf

#### TCP_WRAPPERS CONFIGURATIONS (was GEN006620)
cp -f ./config/hosts.allow /etc/hosts.allow
cp -f ./config/hosts.deny /etc/hosts.deny

if [ -z "$QUIET" ]; then
	echo "Done." | tee -a $LOG
else
	echo "Done." >> $LOG
fi

#### CLEAN TEMP FILES ON REBOOT WITH SCRUB (Server, Workstation)
rpm -q scrub &>/dev/null
if [ $? -eq 0 ]; then
	cp ./config/clean_system /etc/init.d/clean_system
	/sbin/chkconfig --add clean_system
	/sbin/chkconfig --level 06 clean_system on
	/sbin/chkconfig --level 12345 clean_system off
else
	if [ -z "$QUIET" ]; then
		echo
		echo "Scrub not installed. Secure /tmp and /var/tmp wipe service not installed." | tee -a $LOG
	else
		echo
		echo "Scrub not installed. Secure /tmp and /var/tmp wipe service not installed." >> $LOG
	fi
fi

# SECURITY ISSUES
echo >> $LOG
for i in `ls scripts/*.sh`; do 
	if [ -z "$QUIET" ]; then
		echo  "#### Executing Script: $i" | tee -a $LOG
		sh $i 2>&1 | tee -a $LOG
	else
		echo "#### Executing Script: $i" >> $LOG
		sh $i >> $LOG
	fi
done;

# CUSTOM HARDENING
if [ -z "$QUIET" ]; then
	echo
	echo -e "\033[1mAdditional Hardening\033[0m"
	echo
fi
echo >> $LOG
echo "Additional Hardening Scripts" >> $LOG
echo >> $LOG
for i in `ls misc/*.sh`; do
	if [ -z "$QUIET" ]; then
		echo  "#### Executing Script: $i" | tee -a $LOG
		sh $i 2>&1 | tee -a $LOG
	else
		echo "#### Executing Script: $i" >> $LOG
		sh $i >> $LOG
	fi
done;

if [ -z "$QUIET" ]; then
	echo 2>&1 | tee -a $LOG;
	tput setaf 2;echo -e "\033[1mConfiguration Complete!\033[0m";tput sgr0
fi
echo >> $LOG
echo "Configuration Complete!" >> $LOG

exit 0
